### DISCUSSION TITLE
Convenient `diff` of Jupyter Notebooks in PyCharm

### DISCUSSION THREAD

**Discussion by xandrd** [spedas/pyspedas#985] [General]
Hello everyone,

I'd like to share a solution for more convenient analysis of differences (e.g., `diff`) in Jupyter Notebooks using PyCharm. The default diff tool in PyCharm compares notebooks by their raw JSON representation, which is not convenient. However, there's a third-party tool called [`nbdime`](https://nbdime.readthedocs.io/) that offers a content-aware diff.

I came across a discussion on [Stack Overflow](https://stackoverflow.com/questions/46965182/using-nbdime-for-jupyter-notebooks-under-pycharm) about integrating `nbdime` with PyCharm. I'd like to share the steps for setting it up.

<details>
<summary> TL;DR (short version) </summary>

#### **Installation:**

```bash
pip install nbdime
pip install jupyter-server==2.12.5
```

#### **Configuration:**

1. Navigate to `File -> Preferences -> Tools -> Diff & Merge -> External Diff`.
2. Add `nbdiff-web.exe` (located in `venv\Scripts\nbdiff-web.exe`) with the arguments:  
   `--ignore-details --ignore-metadata --ignore-outputs %1 %2`.
3. Associate `Jupyter` with the new external diff tool.

#### **Usage:**

Right-click on the notebook and select `Git -> Show Diff`.
</details>
---

### Step 1: Install `nbdime`

You can install `nbdime` using your terminal:

```bash
pip install nbdime
```

Alternatively, you can use PyCharm’s package manager:  
![image](https://github.com/user-attachments/assets/112b6986-666f-4125-9bdc-2de1120fe0d2)
> [!IMPORTANT]
> The latest version (`4.0.1`) of `nbdime` has a known [issue](https://github.com/jupyter/nbdime/issues/749) with the Jupyter server. To avoid this, install version `2.12.5` of the Jupyter server:
> ```bash
> pip install jupyter-server==2.12.5
> ```
> ![image](https://github.com/user-attachments/assets/f7552227-2352-448a-b4b6-6985bd554844)

---

### Step 2: Configuring PyCharm

1. In PyCharm, go to `File -> Preferences -> Tools -> Diff & Merge -> External Diff Tools`, and enable external tools.
2. Click `+` to add a new tool, then navigate to your Python environment to find `nbdiff-web`. For Windows, this is usually located in the `Scripts` folder as `nbdiff-web.exe`.
3. Add the following argument pattern (see more details [here](https://nbdime.readthedocs.io/en/latest/cli.html#common-diff-options)):  
   `--ignore-details --ignore-metadata --ignore-outputs %1 %2`.
4. Click `OK`.  
   ![image](https://github.com/user-attachments/assets/3e4b26f1-c499-4def-8838-4803499236aa)

Next, associate the external diff tool with Jupyter notebooks:
- In the relevant section, click `+`, select `Jupyter`, and choose `nbdiff-web` as the diff tool.  
  ![image](https://github.com/user-attachments/assets/2bfbdf1f-a013-4699-a998-decb494f889d)

---

### Step 3: Usage

From now on, when you double-click a Jupyter notebook in the `commit` view, the differences will be displayed in your external browser rather than in PyCharm.  
![image](https://github.com/user-attachments/assets/de814249-0400-415f-904c-cfb83ae8f035)

Alternatively, right-click on a notebook and select `Git -> Show Diff`.  
![image](https://github.com/user-attachments/assets/b2405c2f-75a5-4f98-b500-e091ab2a1bee)

---

I hope this helps streamline your workflow when working with Jupyter notebooks in PyCharm!

**Comment by jameswilburlewis** [comment]
I just tried it....seems to work well!

The latest nbdime release (4.0.2) may have fixed the jupyter-server incompatibility issue.   My installation was pretty old, so I just updated to the latest jupyter-server version 2.14.2, and that seems to work.  I haven't tried any of the older releases.

When you're looking for the nbdiff-web program to update your Pycharm preferences, if you happen to be on a Mac, I found it in venv/bin in the virtual environment I'm using.  Also, instead of File->Preferences->Tools, I had to go to Pycharm->Settings->Tools, or click the gear icon for "IDE and Project Settings" in the upper right and navigate to Tools.
